File: frontend/src/services/apiConfigs.ts
/**
 * API Configuration Service
 * Handles API configuration management for businesses
 */

import { apiClient } from './api';

export interface APIProvider {
  provider: string;
  display_name: string;
  description: string;
  requires_secret: boolean;
  requires_oauth: boolean;
  documentation_url?: string;
}

export interface APIConfig {
  id: string;
  business_id: string;
  provider: string;
  provider_name: string;
  max_tokens_per_month?: number;
  tokens_used_current_month: number;
  rate_limit_per_hour?: number;
  requests_this_hour: number;
  cost_per_1k_tokens?: number;
  estimated_monthly_cost?: number;
  is_active: boolean;
  last_tested_at?: string;
  last_test_status?: string;
  last_test_message?: string;
  api_key_masked?: string;
  has_secret: boolean;
  has_access_token: boolean;
  has_refresh_token: boolean;
  created_at: string;
  updated_at: string;
}

export interface APIConfigCreate {
  business_id: string;
  provider: string;
  provider_name: string;
  api_key?: string;
  api_secret?: string;
  access_token?: string;
  refresh_token?: string;
  max_tokens_per_month?: number;
  rate_limit_per_hour?: number;
  cost_per_1k_tokens?: number;
  is_active?: boolean;
  config_data?: string;
}

export interface APIConfigUpdate {
  provider_name?: string;
  api_key?: string;
  api_secret?: string;
  access_token?: string;
  refresh_token?: string;
  max_tokens_per_month?: number;
  rate_limit_per_hour?: number;
  cost_per_1k_tokens?: number;
  is_active?: boolean;
  config_data?: string;
}

export interface APITestResult {
  success: boolean;
  status: string;
  message: string;
  response_time_ms?: number;
  tested_at: string;
}

export interface APIUsageStats {
  config_id: string;
  provider: string;
  tokens_used_current_month: number;
  max_tokens_per_month?: number;
  requests_this_hour: number;
  rate_limit_per_hour?: number;
  estimated_monthly_cost?: number;
  usage_percentage?: number;
}

/**
 * Get list of available API providers
 */
export const getAvailableProviders = async (): Promise<APIProvider[]> => {
  const response = await apiClient.get('/api/v1/api-configs/providers');
  return response.data;
};

/**
 * Get all API configurations for a business
 */
export const getAPIConfigs = async (
  businessId: string,
  provider?: string,
  isActive?: boolean
): Promise<APIConfig[]> => {
  const params: any = { business_id: businessId };
  if (provider) params.provider = provider;
  if (isActive !== undefined) params.is_active = isActive;

  const response = await apiClient.get('/api/v1/api-configs', { params });
  return response.data;
};

/**
 * Get a specific API configuration
 */
export const getAPIConfig = async (
  configId: string,
  businessId: string
): Promise<APIConfig> => {
  const response = await apiClient.get(`/api/v1/api-configs/${configId}`, {
    params: { business_id: businessId }
  });
  return response.data;
};

/**
 * Create a new API configuration
 */
export const createAPIConfig = async (
  config: APIConfigCreate
): Promise<APIConfig> => {
  const response = await apiClient.post('/api/v1/api-configs', config);
  return response.data;
};

/**
 * Update an existing API configuration
 */
export const updateAPIConfig = async (
  configId: string,
  businessId: string,
  update: APIConfigUpdate
): Promise<APIConfig> => {
  const response = await apiClient.put(`/api/v1/api-configs/${configId}`, update, {
    params: { business_id: businessId }
  });
  return response.data;
};

/**
 * Delete an API configuration
 */
export const deleteAPIConfig = async (
  configId: string,
  businessId: string
): Promise<void> => {
  await apiClient.delete(`/api/v1/api-configs/${configId}`, {
    params: { business_id: businessId }
  });
};

/**
 * Test API connection
 */
export const testAPIConnection = async (
  configId: string,
  businessId: string
): Promise<APITestResult> => {
  const response = await apiClient.post(`/api/v1/api-configs/${configId}/test`, null, {
    params: { business_id: businessId }
  });
  return response.data;
};

/**
 * Get usage statistics for all API configurations
 */
export const getUsageStats = async (
  businessId: string
): Promise<APIUsageStats[]> => {
  const response = await apiClient.get('/api/v1/api-configs/usage/stats', {
    params: { business_id: businessId }
  });
  return response.data;
};

/**
 * Record API usage (internal)
 */
export const recordAPIUsage = async (
  configId: string,
  businessId: string,
  tokensUsed: number = 1
): Promise<void> => {
  await apiClient.post(`/api/v1/api-configs/${configId}/usage`, null, {
    params: { business_id: businessId, tokens_used: tokensUsed }
  });
};

/**
 * Reset monthly usage counters
 */
export const resetMonthlyUsage = async (
  businessId: string
): Promise<{ message: string }> => {
  const response = await apiClient.post('/api/v1/api-configs/usage/reset/monthly', null, {
    params: { business_id: businessId }
  });
  return response.data;
};

/**
 * Reset hourly usage counters
 */
export const resetHourlyUsage = async (
  businessId: string
): Promise<{ message: string }> => {
  const response = await apiClient.post('/api/v1/api-configs/usage/reset/hourly', null, {
    params: { business_id: businessId }
  });
  return response.data;
};

File: backend/app/api/routes/api_configs.py
"""
API Configuration Routes
CRUD operations and testing for API configurations
"""
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user, get_db
from app.models.user import User, UserRole
from app.schemas.api_config import (
    APIConfigCreate,
    APIConfigUpdate,
    APIConfigResponse,
    APIConfigTestResponse,
    APIUsageStats,
    APIProviderInfo,
    AVAILABLE_PROVIDERS,
)
from app.services.api_config_service import APIConfigService
from app.services.api_tester import APITester


router = APIRouter()


def check_business_access(current_user: User, business_id: UUID) -> None:
    """Check if user has access to the business"""
    if current_user.role == UserRole.SYSTEM_ADMIN:
        return  # System admin can access any business

    if current_user.business_id != business_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have access to this business"
        )


@router.get("/providers", response_model=List[APIProviderInfo])
async def list_available_providers():
    """
    Get list of available API providers (public endpoint - no auth required)
    """
    return AVAILABLE_PROVIDERS


@router.get("", response_model=List[APIConfigResponse])
async def list_api_configs(
    business_id: UUID = Query(..., description="Business ID"),
    provider: Optional[str] = Query(None, description="Filter by provider"),
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Get list of API configurations for a business

    - **business_id**: Business UUID
    - **provider**: Optional provider filter
    - **is_active**: Optional active status filter
    """
    check_business_access(current_user, business_id)

    configs, total = APIConfigService.get_api_configs(
        db=db,
        business_id=business_id,
        skip=skip,
        limit=limit,
        provider=provider,
        is_active=is_active
    )

    # Convert to response schema (masks sensitive data)
    return [config.to_dict(include_sensitive=False) for config in configs]


@router.get("/{config_id}", response_model=APIConfigResponse)
async def get_api_config(
    config_id: UUID,
    business_id: UUID = Query(..., description="Business ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Get a specific API configuration

    - **config_id**: API Configuration UUID
    - **business_id**: Business UUID
    """
    check_business_access(current_user, business_id)

    config = APIConfigService.get_api_config_by_id(db, config_id, business_id)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API configuration not found"
        )

    return config.to_dict(include_sensitive=False)


@router.post("", response_model=APIConfigResponse, status_code=status.HTTP_201_CREATED)
async def create_api_config(
    config_create: APIConfigCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Create a new API configuration

    - **business_id**: Business UUID
    - **provider**: API provider (newsapi, openai, google_search, etc.)
    - **api_key**: API key (will be encrypted)
    - **api_secret**: Optional API secret (will be encrypted)
    """
    check_business_access(current_user, config_create.business_id)

    try:
        config = APIConfigService.create_api_config(
            db=db,
            config_create=config_create,
            user_id=current_user.id
        )
        return config.to_dict(include_sensitive=False)

    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create API configuration: {str(e)}"
        )


@router.put("/{config_id}", response_model=APIConfigResponse)
async def update_api_config(
    config_id: UUID,
    config_update: APIConfigUpdate,
    business_id: UUID = Query(..., description="Business ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Update an existing API configuration

    - **config_id**: API Configuration UUID
    - **business_id**: Business UUID
    """
    check_business_access(current_user, business_id)

    config = APIConfigService.update_api_config(
        db=db,
        config_id=config_id,
        business_id=business_id,
        config_update=config_update,
        user_id=current_user.id
    )

    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API configuration not found"
        )

    return config.to_dict(include_sensitive=False)


@router.delete("/{config_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_api_config(
    config_id: UUID,
    business_id: UUID = Query(..., description="Business ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Delete an API configuration

    - **config_id**: API Configuration UUID
    - **business_id**: Business UUID
    """
    check_business_access(current_user, business_id)

    # Only business admin or system admin can delete
    if current_user.role not in [UserRole.SYSTEM_ADMIN, UserRole.BUSINESS_ADMIN]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only administrators can delete API configurations"
        )

    success = APIConfigService.delete_api_config(db, config_id, business_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API configuration not found"
        )


@router.post("/{config_id}/test", response_model=APIConfigTestResponse)
async def test_api_connection(
    config_id: UUID,
    business_id: UUID = Query(..., description="Business ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Test API connection for a configuration

    - **config_id**: API Configuration UUID
    - **business_id**: Business UUID
    """
    check_business_access(current_user, business_id)

    config = APIConfigService.get_api_config_by_id(db, config_id, business_id)
    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API configuration not found"
        )

    result = await APITester.test_connection(db, config)
    return result


@router.get("/usage/stats", response_model=List[APIUsageStats])
async def get_usage_stats(
    business_id: UUID = Query(..., description="Business ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Get usage statistics for all API configurations

    - **business_id**: Business UUID
    """
    check_business_access(current_user, business_id)

    stats = APIConfigService.get_usage_stats(db, business_id)
    return stats


@router.post("/{config_id}/usage", status_code=status.HTTP_204_NO_CONTENT)
async def record_api_usage(
    config_id: UUID,
    business_id: UUID = Query(..., description="Business ID"),
    tokens_used: int = Query(1, ge=1, description="Number of tokens used"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Record API usage (internal endpoint)

    - **config_id**: API Configuration UUID
    - **business_id**: Business UUID
    - **tokens_used**: Number of tokens/requests used
    """
    check_business_access(current_user, business_id)

    config = APIConfigService.record_api_usage(
        db=db,
        config_id=config_id,
        business_id=business_id,
        tokens_used=tokens_used
    )

    if not config:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API configuration not found"
        )


@router.post("/usage/reset/monthly", status_code=status.HTTP_200_OK)
async def reset_monthly_usage(
    business_id: UUID = Query(..., description="Business ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Reset monthly usage counters (admin only)

    - **business_id**: Business UUID
    """
    check_business_access(current_user, business_id)

    # Only admins can reset usage
    if current_user.role not in [UserRole.SYSTEM_ADMIN, UserRole.BUSINESS_ADMIN]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only administrators can reset usage counters"
        )

    count = APIConfigService.reset_monthly_usage(db, business_id)
    return {"message": f"Reset usage for {count} configurations"}


@router.post("/usage/reset/hourly", status_code=status.HTTP_200_OK)
async def reset_hourly_usage(
    business_id: UUID = Query(..., description="Business ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Reset hourly usage counters (admin only)

    - **business_id**: Business UUID
    """
    check_business_access(current_user, business_id)

    # Only admins can reset usage
    if current_user.role not in [UserRole.SYSTEM_ADMIN, UserRole.BUSINESS_ADMIN]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only administrators can reset usage counters"
        )

    count = APIConfigService.reset_hourly_usage(db, business_id)
    return {"message": f"Reset hourly usage for {count} configurations"}
